util/topsql/reporter/datamodel.go

```diff
@@ -0,0 +1,685 @@
+//
+// collecting:
+//     - records: { sqlPlanDigest => record | sqlPlanDigest => record | ... }
+//     - evicted: { sqlPlanDigest | sqlPlanDigest | ... }
+//
+// cpuRecords: [ tracecpu.SQLCPUTimeRecord | tracecpu.SQLCPUTimeRecord | ... ]
+//
+// normalizeSQLMap: { sqlDigest => normalizedSQL | sqlDigest => normalizedSQL | ... }
+//
+// normalizePlanMap: { planDigest => normalizedPlan | planDigest => normalizedPlan | ... }
+
+// keyOthers is the key to store the aggregation of all records that is out of Top N.
+const keyOthers = ""
+
+// tsItem is a self-contained complete piece of data for a certain timestamp.
+type tsItem struct {
+	timestamp uint64
+	cpuTimeMs uint32
+	stmtStats stmtstats.StatementStatsItem
+}
+
+func zeroTsItem() tsItem {
+	return tsItem{
+		stmtStats: stmtstats.StatementStatsItem{
+			KvStatsItem: stmtstats.KvStatementStatsItem{
+				KvExecCount: map[string]uint64{},
+			},
+		},
+	}
+}
+
+// toProto converts the tsItem to the corresponding protobuf representation.
+func (i *tsItem) toProto() *tipb.TopSQLRecordItem {
+	return &tipb.TopSQLRecordItem{
+		TimestampSec:    i.timestamp,
+		CpuTimeMs:       i.cpuTimeMs,
+		StmtExecCount:   i.stmtStats.ExecCount,
+		StmtKvExecCount: i.stmtStats.KvStatsItem.KvExecCount,
+		// TODO: add duration
+		// StmtDurationSumNs: i.stmtStats.DurationSumNs,
+		// Convert more indicators here.
+	}
+}
+
+var _ sort.Interface = &tsItems{}
+
+// tsItems is a sortable list of tsItem, sort by tsItem.timestamp (asc).
+type tsItems []tsItem
+
+func (ts tsItems) Len() int {
+	return len(ts)
+}
+
+func (ts tsItems) Less(i, j int) bool {
+	return ts[i].timestamp < ts[j].timestamp
+}
+
+func (ts tsItems) Swap(i, j int) {
+	ts[i], ts[j] = ts[j], ts[i]
+}
+
+func (ts tsItems) sorted() bool {
+	for n := 0; n < len(ts)-1; n++ {
+		if ts[n].timestamp > ts[n+1].timestamp {
+			return false
+		}
+	}
+	return true
+}
+
+// toProto converts the tsItems to the corresponding protobuf representation.
+func (ts tsItems) toProto() []*tipb.TopSQLRecordItem {
+	capacity := len(ts)
+	if capacity == 0 {
+		return nil
+	}
+	items := make([]*tipb.TopSQLRecordItem, 0, capacity)
+	for _, i := range ts {
+		items = append(items, i.toProto())
+	}
+	return items
+}
+
+var _ sort.Interface = &record{}
+
+// record represents the cumulative tsItem in current minute window.
+// record do not guarantee the tsItems is sorted by timestamp when there is a time jump backward.
+// record is also sortable, and the tsIndex will be updated while sorting the internal tsItems.
+type record struct {
+	sqlDigest      []byte
+	planDigest     []byte
+	totalCPUTimeMs uint64
+	tsItems        tsItems
+
+	// tsIndex is used to quickly find the corresponding tsItems index through timestamp.
+	tsIndex map[uint64]int // timestamp => index of tsItems
+}
+
+func newRecord(sqlDigest, planDigest []byte) *record {
+	listCap := variable.TopSQLVariable.ReportIntervalSeconds.Load()/variable.TopSQLVariable.PrecisionSeconds.Load() + 1
```

Comment:
Prefer to add a upper clamp, otherwise setting the variable incorrectly will result in OOM, which is horrible.

Reply:
done